crates/aesgcm/src/core/aesgcm.rs

```diff
@@ -1,78 +1,119 @@
-use aes_gcm::{
-    aead::{consts::U12, generic_array::GenericArray, Aead, KeyInit},
-    Aes256Gcm, Nonce,
+use cloudproof_cover_crypt::reexport::crypto_core::{
+    Aes256Gcm as Aes256GcmRust, CryptoCoreError, Dem, FixedSizeCBytes, Instantiable, Nonce,
+    SymmetricKey,
 };
 
 use crate::error::AesGcmError;
 
-pub const KEY_LENGTH: usize = 32;
-pub const NONCE_LENGTH: usize = 12;
-
 pub const BLOCK_LENGTH: usize = 16;
 
-/// The `ReExposedAesGcm` struct contains an instance of the `Aes256Gcm` cipher
-/// and a nonce of length 12. The crate `aesgcm` has received one security audit
-/// by NCC Group, with no significant findings. We would like to thank
-/// `MobileCoin` for funding the audit.
+/// The `encrypt` function takes a key, nonce, plaintext, and authenticated data
+/// as input and returns the encrypted ciphertext using AES-256 GCM encryption
+/// in Rust.
+///
+/// Arguments:
 ///
-/// Properties:
+/// * `key`: The `key` parameter is a 32-byte array representing the encryption
+///   key
+/// used in the AES-256-GCM encryption algorithm.
+/// * `nonce`: The `nonce` parameter is a unique value used as an input to the
+/// encryption algorithm. It is typically a random value that should never be
+/// reused with the same key. The length of the `nonce` is determined by the
+/// `Aes256GcmRust::NONCE_LENGTH` constant,
+/// * `plaintext`: The `plaintext` parameter is the data that you want to
+///   encrypt.
+/// It is a slice of bytes (`&[u8]`) that represents the raw data that you want
+/// to protect.
+/// * `authenticated_data`: The `authenticated_data` parameter is additional
+///   data
+/// that is authenticated but not encrypted. It is used to provide integrity and
+/// authenticity to the encrypted data. This data is included in the
+/// authentication tag calculation but is not included in the encrypted output.
 ///
-/// * `nonce`: The `nonce` property is a 12-byte array used as a unique value
-///   for each encryption operation in the AES-GCM encryption mode. It is
-///   important that the nonce is never reused with the same key, as this can
-///   compromise the security of the encryption.
-pub struct ReExposedAesGcm {
-    cipher: Aes256Gcm,
-    nonce: GenericArray<u8, U12>,
+/// Returns:
+///
+/// a `Result` type, specifically `Result<Vec<u8>, AesGcmError>`. This means
+/// that the function can either return an `Ok` variant containing a `Vec<u8>`
+/// (the encrypted data) or an `Err` variant containing an `AesGcmError` (an
+/// error occurred during encryption).
+pub fn encrypt(
+    key: [u8; Aes256GcmRust::KEY_LENGTH],
+    nonce: [u8; Aes256GcmRust::NONCE_LENGTH],
+    plaintext: &[u8],
+    authenticated_data: &[u8],
+) -> Result<Vec<u8>, AesGcmError> {
+    let key = SymmetricKey::try_from_bytes(key)?;
+    let nonce = Nonce::try_from_bytes(nonce)?;
+    Ok(Aes256GcmRust::new(&key).encrypt(&nonce, plaintext, Some(authenticated_data))?)
```

Comment:
Why don't you manage type conversion in the interfaces? I am not sure this wrapper is of much use.

Reply:
good point. Done here https://github.com/Cosmian/cloudproof_rust/pull/34/commits/83d5042e0a5c35c8c6962b6e310c1a42fe964220